app/lib/cluster_viz_service.rb

```diff
@@ -174,47 +174,64 @@ def self.compute_aspect_ratios(range)
     aspect
   end
 
-  # set the range for a plotly scatter, will default to data-defined if cluster hasn't defined its own ranges
-  # dynamically determines range based on inputs & available axes
-  def self.set_range(cluster, inputs)
-    # select coordinate axes from inputs
-    domain_keys = inputs.map(&:keys).flatten.uniq.select {|i| [:x, :y, :z].include?(i)}
-    range = Hash[domain_keys.zip]
-    if cluster.has_range?
-      # use study-provided range if available
-      range = cluster.domain_ranges
-    else
-      # take the minmax of each domain across all groups, then the global minmax
-      raw_values = []
-      inputs.each do |input|
-        domain_keys.each do |domain|
-          domain_range = RequestUtils.get_minmax(input[domain])
-          # RequestUtils.get_minmax will discard NaN/nil values that were ingested
-          # only add domain range to list if we have a valid minmax
-          raw_values << domain_range if domain_range.any?
-        end
-      end
-      aggregate_range = raw_values.flatten.minmax
-      # add 2% padding to range
-      padding = (aggregate_range.first - aggregate_range.last) * 0.02
-      absolute_range = [aggregate_range.first + padding, aggregate_range.last - padding]
-      range[:x] = absolute_range
-      range[:y] = absolute_range
-      range[:z] = absolute_range
-    end
-    range
-  end
-
   # generic method to populate data structure to render a cluster scatter plot
   # uses cluster_group model and loads annotation for both group & numeric plots
   # data values are pulled from associated data_array entries for each axis and annotation/text value
-  def self.load_cluster_group_data_array_points(study, cluster, annotation, subsample_threshold=nil, colorscale=nil)
+  def self.load_cluster_group_data_array_points(study, cluster, annotation, subsample_threshold=nil, include_coords: true, include_cells: true, include_annotations: true)
     # construct annotation key to load subsample data_arrays if needed, will be identical to params[:annotation]
     subsample_annotation = "#{annotation[:name]}--#{annotation[:type]}--#{annotation[:scope]}"
-    x_array = cluster.concatenate_data_arrays('x', 'coordinates', subsample_threshold, subsample_annotation)
-    y_array = cluster.concatenate_data_arrays('y', 'coordinates', subsample_threshold, subsample_annotation)
-    z_array = cluster.concatenate_data_arrays('z', 'coordinates', subsample_threshold, subsample_annotation)
-    cells = cluster.concatenate_data_arrays('text', 'cells', subsample_threshold, subsample_annotation)
+
+    x_array = []
+    y_array = []
+    z_array = []
+    cells = []
+    annotation_array = []
+
+    if annotation[:scope] == 'user'
+      user_annotation = UserAnnotation.find(annotation[:id])
+      subsample_annotation = user_annotation.formatted_annotation_identifier
+      if include_coords
```

Comment:
I think this can all be simplified:
```
data_source = cluster
if annotation[:scope] == 'user'
  user_annotation = UserAnnotation.find(annotation[:id])
  subsample_annotation = user_annotation.formatted_annotation_identifier
  data_source = user_annotation
end
if include_coords
  x_array = data_source.concatenate_user_data_arrays('x', 'coordinates', subsample_threshold, subsample_annotation)
  y_array = data_source.concatenate_user_data_arrays('y', 'coordinates', subsample_threshold, subsample_annotation)
  z_array = data_source.concatenate_user_data_arrays('z', 'coordinates', subsample_threshold, subsample_annotation)
end
if include_cells || annotation[:scope] == 'study'
  # for study annotations, we have to grab the cluster cells to match with the study-wide annotation
  cells = data_source.concatenate_data_arrays('text', 'cells', subsample_threshold, subsample_annotation)
end
```

Reply:
NICE!